extension/src/lib.rs

```diff
@@ -0,0 +1,295 @@
+        window_size: Milliseconds, // the size of a window to delta over
+        time: TimestampTz,
+        val: f64,
+    ) -> Option<Pox<GapfillDeltaTransition>> {
+        if time < lowest_time || time > greatest_time {
+            elog!(Error, "input time less than lowest time")
+        }
+
+        let mut state = state.unwrap_or_else(|| {
+            let state: Pox<_> = GapfillDeltaTransition::new(
+                lowest_time,
+                greatest_time,
+                window_size,
+                step_size,
+                true,
+                false,
+            ).into();
+            state
+        });
+
+        state.add_data_point(time, val);
+
+        Some(state)
+    }
+
+    pub fn gapfill_delta_final(state: Option<Pox<GapfillDeltaTransition>>) -> Option<*mut ArrayType> {
+        state.map(|mut s| s.to_pg_array())
+    }
+}
+
+#[derive(Debug)]
+struct GapfillDeltaTransition {
+    window: VecDeque<(TimestampTz, f64)>,
+    deltas: Vec<Datum>,
+    nulls: Vec<bool>,
+    current_window_max: TimestampTz,
+    current_window_min: TimestampTz,
+    step_size: TimestampTz,
+    greatest_time: TimestampTz,
+    is_counter: bool,
+    is_rate: bool,
+}
+
+impl GapfillDeltaTransition {
+    pub fn new(
+        lowest_time: TimestampTz,
+        greatest_time: TimestampTz,
+        window_size: Milliseconds,
+        step_size: Milliseconds,
+        is_counter: bool,
+        is_rate: bool
+    ) -> Self {
+        let mut expected_deltas = (greatest_time - lowest_time) / (step_size * USECS_PER_MS);
+        if (greatest_time - lowest_time) % (step_size * USECS_PER_MS) != 0 {
+            expected_deltas += 1
+        }
+        GapfillDeltaTransition{
+            window: VecDeque::default(),
+            deltas: Vec::with_capacity(expected_deltas as usize),
+            nulls: Vec::with_capacity(expected_deltas as usize),
+            current_window_max: lowest_time + window_size*USECS_PER_MS,
+            current_window_min: lowest_time,
+            step_size: step_size*USECS_PER_MS,
+            greatest_time,
+            is_counter,
+            is_rate,
+        }
+    }
+
+    fn add_data_point(&mut self, time: TimestampTz, val: f64) {
+        // skip stale NaNs
+        const STALE_NAN: u64 = 0x7ff0000000000002;
+        if val.to_bits() == STALE_NAN {
+            return
+        }
+
+        while !self.in_current_window(time) {
+            self.flush_current_window()
+        }
+
+        self.window.push_back((time, val));
+    }
+
+    fn in_current_window(&self, time: TimestampTz) -> bool {
+        time <= self.current_window_max
+    }
+
+    fn flush_current_window(&mut self) {
+        self.add_delta_for_current_window();
+
+        self.current_window_min += self.step_size;
+        self.current_window_max += self.step_size;
+
+        let current_window_min = self.current_window_min;
+        while self.window.front().map(|(time, _)| *time < current_window_min).unwrap_or(false) {
```

Comment:
not sure I follow this ... can discuss tomorrow during walkthru

Reply:
if's basically
```python
while window.front() exists and window.front() < current_window_min
```